client/comms/wsconn.go

```diff
@@ -0,0 +1,320 @@
+package comms
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/decred/dcrdex/server/comms/msgjson"
+	"github.com/gorilla/websocket"
+)
+
+const (
+	// bufferSize is buffer size for the websocket connection's read channel.
+	readBuffSize = 128
+)
+
+// WsCfg is the configuration struct for initializing a WsConn.
+type WsCfg struct {
+	// The websocket host.
+	Host string
+	// The websocket api path.
+	Path string
+	// The maximum time in seconds to wait for a ping from the server.
+	PingWait time.Duration
+	// The maximum time in seconds to write to the connection.
+	WriteWait time.Duration
+	// The rpc certificate file path.
+	RpcCert string
+	// ReconnectSync runs the needed reconnection synchronisation after
+	// a disconnect.
+	ReconnectSync func()
+	// The dex client context.
+	Ctx context.Context
+	// The dex client context cancel func.
+	Cancel context.CancelFunc
+}
+
+// WsConn represents a client websocket connection.
+type WsConn struct {
+	reconnects   uint64
+	rID          uint64
+	cfg          *WsCfg
+	ws           *websocket.Conn
+	wsMtx        sync.Mutex
+	tlsCfg       *tls.Config
+	readCh       chan *msgjson.Message
+	sendCh       chan *msgjson.Message
+	reconnectCh  chan struct{}
+	req          map[uint64]*msgjson.Message
+	reqMtx       sync.RWMutex
+	connected    bool
+	connectedMtx sync.RWMutex
+	once         sync.Once
+}
+
+// filesExists reports whether the named file or directory exists.
+func fileExists(name string) bool {
+	_, err := os.Stat(name)
+	return !os.IsNotExist(err)
+}
+
+// NewWsConn creates a client websocket connection.
+func NewWsConn(cfg *WsCfg) (*WsConn, error) {
+	if cfg.PingWait < 0 {
+		return nil, fmt.Errorf("ping wait cannot be negative")
+	}
+
+	if cfg.WriteWait < 0 {
+		return nil, fmt.Errorf("write wait cannot be negative")
+	}
+
+	var tlsConfig *tls.Config
+	if fileExists(cfg.RpcCert) {
+		rootCAs, _ := x509.SystemCertPool()
+		if rootCAs == nil {
+			rootCAs = x509.NewCertPool()
+		}
+
+		certs, err := ioutil.ReadFile(cfg.RpcCert)
+		if err != nil {
+			return nil, fmt.Errorf("file reading error: %v", err)
+		}
+
+		if ok := rootCAs.AppendCertsFromPEM(certs); !ok {
+			return nil, fmt.Errorf("unable to append cert")
+		}
+
+		tlsConfig = &tls.Config{
+			RootCAs:            rootCAs,
+			InsecureSkipVerify: true,
```

Comment:
I don't understand.  If the certs are specified, they should be verified.  Essentially InsecureSkipVerify says don't bother looking at the root CAs, right?

Reply:
Yup, I'm reverting back to having the option avialable via the config for tests. Good catch.